File: Lumi/Core/Services/DifficultyManager.swift
import Foundation
import SwiftData

/// Manages adaptive difficulty based on performance
@Observable
final class DifficultyManager {
    private var modelContext: ModelContext?

    // Track consecutive answers per activity type (in-memory for current session)
    private var consecutiveCorrect: [String: Int] = [:]
    private var consecutiveIncorrect: [String: Int] = [:]

    // Current difficulty per activity type
    private var currentDifficulty: [String: Int] = [:]

    static let minDifficulty = 1
    static let maxDifficulty = 4
    static let correctToLevelUp = 3
    static let incorrectToLevelDown = 2

    func configure(with context: ModelContext) {
        self.modelContext = context
        loadProgress()
    }

    private func loadProgress() {
        guard let context = modelContext else { return }

        let descriptor = FetchDescriptor<ActivityProgress>()

        do {
            let results = try context.fetch(descriptor)
            for progress in results {
                currentDifficulty[progress.activityType] = progress.currentDifficulty
                consecutiveCorrect[progress.activityType] = progress.consecutiveCorrect
                consecutiveIncorrect[progress.activityType] = progress.consecutiveIncorrect
            }
        } catch {
            print("Error loading progress: \(error)")
        }
    }

    func getDifficulty(for activityType: String) -> Int {
        return currentDifficulty[activityType] ?? Self.minDifficulty
    }

    func getDifficulty(for activityType: ProblemType) -> Int {
        return getDifficulty(for: activityType.rawValue)
    }

    /// Records an answer and returns the new difficulty level
    @discardableResult
    func recordAnswer(correct: Bool, for activityType: String) -> Int {
        let currentLevel = currentDifficulty[activityType] ?? Self.minDifficulty

        if correct {
            consecutiveCorrect[activityType, default: 0] += 1
            consecutiveIncorrect[activityType] = 0

            // Level up after 3 correct in a row
            if consecutiveCorrect[activityType, default: 0] >= Self.correctToLevelUp {
                let newLevel = min(currentLevel + 1, Self.maxDifficulty)
                currentDifficulty[activityType] = newLevel
                consecutiveCorrect[activityType] = 0
                saveProgress(activityType: activityType)
                return newLevel
            }
        } else {
            consecutiveIncorrect[activityType, default: 0] += 1
            consecutiveCorrect[activityType] = 0

            // Level down after 2 incorrect in a row
            if consecutiveIncorrect[activityType, default: 0] >= Self.incorrectToLevelDown {
                let newLevel = max(currentLevel - 1, Self.minDifficulty)
                currentDifficulty[activityType] = newLevel
                consecutiveIncorrect[activityType] = 0
                saveProgress(activityType: activityType)
                return newLevel
            }
        }

        saveProgress(activityType: activityType)
        return currentLevel
    }

    @discardableResult
    func recordAnswer(correct: Bool, for activityType: ProblemType) -> Int {
        return recordAnswer(correct: correct, for: activityType.rawValue)
    }

    private func saveProgress(activityType: String) {
        guard let context = modelContext else { return }

        let descriptor = FetchDescriptor<ActivityProgress>(
            predicate: #Predicate { $0.activityType == activityType }
        )

        do {
            let results = try context.fetch(descriptor)
            let progress: ActivityProgress

            if let existing = results.first {
                progress = existing
            } else {
                progress = ActivityProgress(activityType: activityType)
                context.insert(progress)
            }

            progress.currentDifficulty = currentDifficulty[activityType] ?? Self.minDifficulty
            progress.consecutiveCorrect = consecutiveCorrect[activityType] ?? 0
            progress.consecutiveIncorrect = consecutiveIncorrect[activityType] ?? 0
            progress.lastPlayedAt = Date()

            try context.save()
        } catch {
            print("Error saving progress: \(error)")
        }
    }

    func recordProblemAttempt(correct: Bool, for activityType: String) {
        guard let context = modelContext else { return }

        let descriptor = FetchDescriptor<ActivityProgress>(
            predicate: #Predicate { $0.activityType == activityType }
        )

        do {
            let results = try context.fetch(descriptor)
            let progress: ActivityProgress

            if let existing = results.first {
                progress = existing
            } else {
                progress = ActivityProgress(activityType: activityType)
                context.insert(progress)
            }

            progress.problemsAttempted += 1
            if correct {
                progress.problemsCorrect += 1
            }

            try context.save()
        } catch {
            print("Error recording problem attempt: \(error)")
        }
    }
}

File: src/lib/services/difficulty-manager.ts
/**
 * Difficulty Manager
 *
 * Manages adaptive difficulty based on performance.
 * Level UP: After 3 consecutive correct answers
 * Level DOWN: After 2 consecutive incorrect answers
 */

import type { ProblemType, DifficultyLevel, ActivityProgress } from '$lib/types'
import { createDefaultActivityProgress } from '$lib/types'

export const MIN_DIFFICULTY: DifficultyLevel = 1
export const MAX_DIFFICULTY: DifficultyLevel = 4
export const CORRECT_TO_LEVEL_UP = 3
export const INCORRECT_TO_LEVEL_DOWN = 2

export class DifficultyManager {
	private progress: Map<ProblemType, ActivityProgress> = new Map()
	private defaultStartingDifficulty: DifficultyLevel = MIN_DIFFICULTY

	constructor() {
		this.initializeProgress()
	}

	/**
	 * Set the default starting difficulty for new problem types.
	 * Used by age service to adjust difficulty based on child's age.
	 * Also updates existing problem types that haven't been played yet.
	 */
	setDefaultStartingDifficulty(difficulty: DifficultyLevel): void {
		this.defaultStartingDifficulty = difficulty

		// Update all problem types that haven't been played yet
		for (const progress of this.progress.values()) {
			if (progress.problemsAttempted === 0) {
				progress.currentDifficulty = difficulty
			}
		}
	}

	/**
	 * Get the current default starting difficulty
	 */
	getDefaultStartingDifficulty(): DifficultyLevel {
		return this.defaultStartingDifficulty
	}

	private initializeProgress(): void {
		const types: ProblemType[] = [
			// Math
			'counting',
			'addition',
			'subtraction',
			'comparison',
			'patterns',
			// Grammar
			'letter-recognition',
			'alphabet-order',
			'initial-letter',
			'word-completion'
		]
		for (const type of types) {
			this.progress.set(type, createDefaultActivityProgress(type, this.defaultStartingDifficulty))
		}
	}

	/**
	 * Load progress from storage
	 */
	loadProgress(data: Record<ProblemType, ActivityProgress>): void {
		for (const [type, progress] of Object.entries(data)) {
			this.progress.set(type as ProblemType, progress)
		}
	}

	/**
	 * Get all progress for persistence
	 */
	getAllProgress(): Record<ProblemType, ActivityProgress> {
		const result: Partial<Record<ProblemType, ActivityProgress>> = {}
		for (const [type, progress] of this.progress) {
			result[type] = progress
		}
		return result as Record<ProblemType, ActivityProgress>
	}

	/**
	 * Get current difficulty for an activity type
	 */
	getDifficulty(activityType: ProblemType): DifficultyLevel {
		return this.progress.get(activityType)?.currentDifficulty ?? MIN_DIFFICULTY
	}

	/**
	 * Get all difficulties as a map (for ProblemService)
	 */
	getAllDifficulties(): Map<ProblemType, DifficultyLevel> {
		const result = new Map<ProblemType, DifficultyLevel>()
		for (const [type, progress] of this.progress) {
			result.set(type, progress.currentDifficulty)
		}
		return result
	}

	/**
	 * Get progress for an activity type
	 */
	getProgress(activityType: ProblemType): ActivityProgress {
		return (
			this.progress.get(activityType) ??
			createDefaultActivityProgress(activityType, this.defaultStartingDifficulty)
		)
	}

	/**
	 * Record an answer and update difficulty
	 * @returns The new difficulty level
	 */
	recordAnswer(correct: boolean, activityType: ProblemType): DifficultyLevel {
		let progress = this.progress.get(activityType)
		if (!progress) {
			progress = createDefaultActivityProgress(activityType, this.defaultStartingDifficulty)
			this.progress.set(activityType, progress)
		}

		// Update attempt counts
		progress.problemsAttempted += 1
		if (correct) {
			progress.problemsCorrect += 1
		}

		if (correct) {
			progress.consecutiveCorrect += 1
			progress.consecutiveIncorrect = 0

			// Level up after 3 correct in a row
			if (progress.consecutiveCorrect >= CORRECT_TO_LEVEL_UP) {
				progress.currentDifficulty = Math.min(
					progress.currentDifficulty + 1,
					MAX_DIFFICULTY
				) as DifficultyLevel
				progress.consecutiveCorrect = 0
			}
		} else {
			progress.consecutiveIncorrect += 1
			progress.consecutiveCorrect = 0

			// Level down after 2 incorrect in a row
			if (progress.consecutiveIncorrect >= INCORRECT_TO_LEVEL_DOWN) {
				progress.currentDifficulty = Math.max(
					progress.currentDifficulty - 1,
					MIN_DIFFICULTY
				) as DifficultyLevel
				progress.consecutiveIncorrect = 0
			}
		}

		return progress.currentDifficulty
	}

	/**
	 * Reset progress for a specific activity type
	 */
	resetProgress(activityType: ProblemType): void {
		this.progress.set(
			activityType,
			createDefaultActivityProgress(activityType, this.defaultStartingDifficulty)
		)
	}

	/**
	 * Reset all progress
	 */
	resetAllProgress(): void {
		this.initializeProgress()
	}

	/**
	 * Get accuracy percentage for an activity type
	 */
	getAccuracy(activityType: ProblemType): number {
		const progress = this.progress.get(activityType)
		if (!progress || progress.problemsAttempted === 0) {
			return 0
		}
		return Math.round((progress.problemsCorrect / progress.problemsAttempted) * 100)
	}
}

// Singleton instance
export const difficultyManager = new DifficultyManager()
